traits/trait_dict_object.py

```diff
@@ -57,18 +50,314 @@ def __init__(self, added=None, changed=None, removed=None):
+        Called to validate a key in the dict.
+        The callable must accept a single key and
+        return a validated key or raise a TraitError
+        If not provided, all keys are considered valid.
+    value_validator : callable, optional
+        Called to validate a value in the dict.
+        The callable must accept a single value and
+        return a validated value or raise a TraitError
+        If not provided, all values are considered valid.
+    notifiers : list, optional
+        A list of callables with the signature::
+
+            notifier(trait_dict, added, changed, removed)
+
+        Where:
+        'removed' is a dict of key-values that are no longer in the dictionary.
+        'added' is a dict of new key-values that have been added.
+        'changed' is a dict with old values previously associated with the key.
+
+    Attributes
+    ----------
+    key_validator : callable
+        Called to validate a key in the dict.
+        The callable must accept a single key and
+        return a validated key or raise a TraitError
+    value_validator : callable
+        Called to validate a value in the dict.
+        The callable must accept a single value and
+        return a validated value or raise a TraitError
+    notifiers : list
+        A list of callables with the signature::
+
+            notifier(trait_dict, added, changed, removed)
+
+        Where:
+        'removed' is a dict of key-values that are no longer in the dictionary.
+        'added' is a dict of new key-values that have been added.
+        'changed' is a dict with old values previously associated with the key.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        self = super().__new__(cls)
+        self.key_validator = accept_anything
+        self.value_validator = accept_anything
+        self.notifiers = []
+        return self
+
+    def __init__(self, value=None, *, key_validator=None,
+                 value_validator=None, notifiers=None):
+
+        if key_validator is not None:
+            self.key_validator = key_validator
+
+        if value_validator is not None:
+            self.value_validator = value_validator
+
+        if notifiers is None:
+            notifiers = []
+        self.notifiers = notifiers
+
+        if value is None:
+            value = {}
+
+        items = value.items() if isinstance(value, dict) else value
+        value = {self.key_validator(key): self.value_validator(value)
+                 for key, value in items}
+
+        super().__init__(value)
+
+    def notify(self, removed, added, changed):
+        """ Call all notifiers.
+
+        This simply calls all notifiers provided by the class, if any.
+        The notifiers are expected to have the signature::
+
+            notifier(trait_dict, removed, added, changed)
+
+        Any return values are ignored.
+        """
+
+        for notifier in self.notifiers:
+            notifier(self, removed=removed, added=added, changed=changed)
```

Comment:
Let's pass by position instead of by keyword here, for consistency with `TraitList` and `TraitSet`.

Note that the way that we make this call matters, and should be documented: externally-supplied notifiers (including those possibly written in C) could potentially need to know whether they're going to be called with positional arguments or keyword arguments.

```suggestion
            notifier(self, removed, added, changed)
```

Reply:
Why do we need to pass `self` to the `notifier`? I notice that we're doing this in `TraitList`, but not in `TraitSet`; we should be consistent one way or the other (so no matter what happens here, we'll need to change one of `TraitList` and `TraitSet`).

Reply:
Opened #1027 to make sure we don't forget this.

Reply:
From discussion in #1028, we want to keep the `self` here (and add it back in `TraitSet`).